### PR TITLE
MMP-32 Create method that accepts the directory and check project str…

### PR DESCRIPTION
**Pull request by WilliamPinho**
…ucture, and create tests

Add application package validator and packaging type and launcher that accepts command line arguments

### CODE REVIEW

File: src/test/java/org/mule/tools/artifact/archiver/api/PackageBuilderTest.java [hunk 1]
```diff
@@ -370,4 +376,132 @@ public void createDeployableFileSettingMulePomFileTest() throws IOException {
...
+        this.packageBuilder.withArchiver(mockedArchiver);
+
+        this.packageBuilder.generateArtifact(targetFolder, mockedDestinationFile);
+
+        verify(appender).doAppend(logCaptor.capture());
+        String SUCCESS_MESSAGE = "File " + mockedDestinationFile.getName() + " has been successfully created";
+        Assert.assertEquals("Info about created destination file not logged", SUCCESS_MESSAGE, logCaptor.getValue().getRenderedMessage());
+    }
+
+    public File[] getFiles() {
```
Comment (damiansima) [inline src/test/java/org/mule/tools/artifact/archiver/api/PackageBuilderTest.java]: this should be private

File: src/test/java/org/mule/tools/artifact/archiver/internal/ApplicationPackageStructureValidatorTest.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        File[] defaultFileStructure = getDefaultFileStructure();
+
+        Assert.assertTrue("This file structure should match the expected structure", applicationPackageStructure.hasExpectedStructure(defaultFileStructure));
+
+        List<File> wrongExpectedStructure = Arrays.asList(defaultFileStructure).subList(0,5);
+
+        Assert.assertFalse("This file structure should not match the expected structure", applicationPackageStructure.hasExpectedStructure(wrongExpectedStructure.toArray(new File[0])));
+    }
+
+    public File[] getDefaultFileStructure() {
```
Comment (damiansima) [inline src/test/java/org/mule/tools/artifact/archiver/internal/ApplicationPackageStructureValidatorTest.java]: this should be private

File: src/test/java/org/mule/tools/artifact/archiver/internal/ApplicationPackageStructureValidatorTest.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        List<File> expectedStructure = new ArrayList<>();
+        File mockedPluginsFolder = mock(File.class);
```
Comment (damiansima) [inline src/test/java/org/mule/tools/artifact/archiver/internal/ApplicationPackageStructureValidatorTest.java]: please follow the convention pluginsFolderMock

File: pom.xml [hunk 4]
```diff
@@ -207,6 +207,17 @@
             <version>${guava.version}</version>
 
         </dependency>
+        <dependency>
+            <groupId>commons-logging</groupId>
+            <artifactId>commons-logging</artifactId>
+            <version>1.2</version>
```
Comment (WilliamPinho) [inline pom.xml]: Done. Also for other dependencies that were already there with without the versions as properties.

File: pom.xml [hunk 5]
```diff
@@ -226,6 +237,13 @@
             <version>${mockito.core.version}</version>
         </dependency>
 
+        <!-- https://mvnrepository.com/artifact/commons-cli/commons-cli -->
```
Comment (WilliamPinho) [inline pom.xml]: Removed.

File: src/main/java/org/mule/tools/artifact/archiver/api/MuleArtifactArchiverLauncher.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        CommandLine cmd = parse(args);
+        String packagingType = cmd.getOptionValue(ARTIFACT_CONTENT_TYPE, PackagingTypeFactory.BINARIES_PACKAGING);
+        PackageBuilder packageBuilder = new PackageBuilder(PackagingTypeFactory.getPackaging(packagingType));
+        String targetFolder = cmd.getOptionValue("target", System.getProperty("user.dir"));
+        if(!cmd.hasOption("packageName")) {
+            System.err.println("The packageName argument is mandatory");
+            System.exit(1);
+        }
+        String destinationFile = cmd.getOptionValue("target", System.getProperty("user.dir"));
+        packageBuilder.generateArtifact(new File(targetFolder), new File(targetFolder + File.separator + destinationFile));
```
Comment (WilliamPinho) [inline src/main/java/org/mule/tools/artifact/archiver/api/MuleArtifactArchiverLauncher.java]: No, it is not correct, my bad. The second one should change to 

String destinationFile = cmd.getOptionValue("packageName");

File: src/main/java/org/mule/tools/artifact/archiver/api/MuleArtifactArchiverLauncher.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        }
+        return null;
+    }
+
+
+    private static void createOptions() {
+        options = new Options();
+        options.addOption(ARTIFACT_CONTENT_TYPE, true, "generate package containing binaries and/or sources");
+        options.addOption("target", true, "root folder where package structure is located");
+        options.addOption("destinationFile", true, "file name for generated file");
```
Comment (WilliamPinho) [inline src/main/java/org/mule/tools/artifact/archiver/api/MuleArtifactArchiverLauncher.java]: Yep! Done.

File: src/main/java/org/mule/tools/artifact/archiver/api/PackageBuilder.java [hunk 8]
```diff
@@ -222,6 +251,30 @@ public MuleArchiver getMuleArchiver() {
         }
         return archiver;
     }
+
+    public void generateArtifact(File targetFolder, File destinationFile) throws IOException {
+        checkMandatoryFolder(targetFolder);
+        checkNotNull(destinationFile);
+        checkArgument(!destinationFile.exists(), "The destination file already exists");
```
Comment (WilliamPinho) [inline src/main/java/org/mule/tools/artifact/archiver/api/PackageBuilder.java]: Yes, we can! Done

File: src/main/java/org/mule/tools/artifact/archiver/api/PackageBuilder.java [hunk 9]
```diff
@@ -222,6 +251,30 @@ public MuleArchiver getMuleArchiver() {
...
+        File[] files = targetFolder.listFiles();
+        if(files == null) {
+            log.warn("The provided target folder is empty");
```
Comment (WilliamPinho) [inline src/main/java/org/mule/tools/artifact/archiver/api/PackageBuilder.java]: Done. Changed the test as well

File: src/main/java/org/mule/tools/artifact/archiver/api/PackageBuilder.java [hunk 10]
```diff
@@ -222,6 +251,30 @@ public MuleArchiver getMuleArchiver() {
...
+        }
+        else if(getApplicationPackageStructureValidator().hasExpectedStructure(files)) {
```
Comment (WilliamPinho) [inline src/main/java/org/mule/tools/artifact/archiver/api/PackageBuilder.java]: Done

File: src/main/java/org/mule/tools/artifact/archiver/internal/PackagingTypeFactory.java [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+                return new BinariesPackaging();
+            case SOURCES_PACKAGING:
+                return new SourcesPackaging();
+            default:
+                return getDefaultPackaging();
+        }
+
+    }
+
+    public static class BinariesAndSourcePackaging implements PackagingType {
```
Comment (WilliamPinho) [inline src/main/java/org/mule/tools/artifact/archiver/internal/PackagingTypeFactory.java]: Done

File: src/test/java/org/mule/tools/artifact/archiver/api/PackageBuilderTest.java [hunk 12]
```diff
@@ -24,25 +30,35 @@
...
+    @Mock AppenderSkeleton appender;
+    @Captor ArgumentCaptor<LoggingEvent> logCaptor;
+
     @Rule
     public TemporaryFolder targetFileFolder = new TemporaryFolder();
 
     @Before
     public void beforeTest() {
         this.packageBuilder = new PackageBuilder();
+        this.mockedDestinationFile = mock(File.class);
```
Comment (WilliamPinho) [inline src/test/java/org/mule/tools/artifact/archiver/api/PackageBuilderTest.java]: Phew! Done

File: src/test/java/org/mule/tools/artifact/archiver/internal/ApplicationPackageStructureValidatorTest.java (hunk 2 above)
Comment (WilliamPinho) [inline src/test/java/org/mule/tools/artifact/archiver/internal/ApplicationPackageStructureValidatorTest.java]: Done

File: src/test/java/org/mule/tools/artifact/archiver/api/PackageBuilderTest.java (hunk 1 above)
Comment (WilliamPinho) [inline src/test/java/org/mule/tools/artifact/archiver/api/PackageBuilderTest.java]: Done

File: src/test/java/org/mule/tools/artifact/archiver/internal/ApplicationPackageStructureValidatorTest.java (hunk 3 above)
Comment (WilliamPinho) [inline src/test/java/org/mule/tools/artifact/archiver/internal/ApplicationPackageStructureValidatorTest.java]: Done

**Pull request merged**